### PR TITLE
feat: Add wasm-bindings and integrate oracle module in app

### PR DESCRIPTION
**Pull request by Tikaryan**
## 1. Overview

- Integrated the Oracle module in `app. go`.

- Added the Stargate whitelisted query `ExchangeRate`  for the contract to fetch the exchange rate.

## 2. Implementation details

- Registered Stargate queries in the wasm keeper.

- Added `Stargate` as a supported feature in the wasm keeper.

- CosmWasm contract will be able to query the keeper using stargate whitelisted queries.

- Currently, we have whitelisted only a single query `persistence.oracle.v1beta1.Query/ExchangeRate` for querying the exchange rate.

## 3. Checklist

- [x] Does the Readme need to be updated?

### CODE REVIEW

File: wasmbindings/message_plugin.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		return &CustomMessenger{
+			wrapped:        old,
+			checkersKeeper: checkersKeeper,
+		}
+	}
+}
+
+type CustomMessenger struct {
+	wrapped        wasmkeeper.Messenger
+	checkersKeeper *oraclekeeper.Keeper
```
Comment (puneet2019) [inline wasmbindings/message_plugin.go:27]: why name it `checkersKeeper`?

**Comment by puneet2019** [comment]
awesome while you are at it, could you also put a sample readme.md or guidelines to add a custom message?
Example I want to add liquid-stake txn, liquid-unstake, be able to query exchange rate.....

File: wasmbindings/message_plugin.go (hunk 1 above)
Comment (Tikaryan) [reply wasmbindings/message_plugin.go]: Updated.

**Review by xlab — APPROVED** [review]
While this initial implementation looks legit, it definitely needs to be compared with Stagate-based queries that use protobuf for encoding and provide more safe routing, which calls the existing queries from the module.

Let's keep custom queries, but once wasm contract is ready then we need to evaluate other option as well.

(see )
https://github.com/CosmWasm/cosmwasm/blob/main/SEMANTICS.md#query-semantics

**Review by xlab — CHANGES_REQUESTED** [review]
meant to request changes.

Also, I agree with @puneet2019 , let's use Osmosis approach with proto-encoded stuff.

File: app/ante_handler.go [hunk 2]
```diff
@@ -44,6 +45,7 @@ func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
 		wasmkeeper.NewCountTXDecorator(options.TXCounterStoreKey),
 		ante.NewRejectExtensionOptionsDecorator(),
 		ante.NewMempoolFeeDecorator(),
+		NewSpamPreventionDecorator(options.OracleKeeper), // spam prevention
```
[the remark this answers is not in the document]
Comment (Tikaryan) [reply app/ante_handler.go]: Done

File: app/app.go [hunk 3]
```diff
@@ -175,6 +180,7 @@ var ModuleAccountPermissions = map[string][]string{
 	lscosmostypes.RewardModuleAccount:        nil,
 	lscosmostypes.UndelegationModuleAccount:  nil,
 	lscosmostypes.RewardBoosterModuleAccount: nil,
+	oracletypes.ModuleName:                   {authtypes.Burner},
```
[the remark this answers is not in the document]
Comment (Tikaryan) [reply app/app.go]: Removed

File: app/spam_prevention.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	"sync"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	oracletypes "github.com/persistenceOne/persistence-sdk/v2/x/oracle/types"
+)
+
+// SpamPreventionDecorator will check if the transaction's gas is smaller than
+// configured hard cap.
+type SpamPreventionDecorator struct {
```
[the remark this answers is not in the document]
Comment (Tikaryan) [reply app/spam_prevention.go]: Removed

File: app/upgrades/v6/untombstoning.go [hunk 5]
```diff
@@ -3,6 +3,7 @@ package v6
 import (
```
[the remark this answers is not in the document]
Comment (Tikaryan) [reply app/upgrades/v6/untombstoning.go]: Removed.

**Comment by Tikaryan** [comment]
> meant to request changes.
> 
> Also, I agree with @puneet2019 , let's use Osmosis approach with proto-encoded stuff.

I think we should make a separate PR for this.

File: app/app.go [hunk 6]
```diff
@@ -574,7 +591,9 @@ func NewApplication(
 
 	// The last arguments can contain custom message handlers, and custom query handlers,
 	// if we want to allow any custom callbacks
-	supportedFeatures := "iterator,staking,stargate"
```
Comment (xlab) [inline app/app.go:577]: Reason to remove `staking` from here?

File: app/app.go [hunk 7]
```diff
@@ -884,13 +909,6 @@ func NewApplication(
 	return app
 }
 
-func (app *Application) CreateUpgradeHandler(mm *module.Manager, configurator module.Configurator) upgradetypes.UpgradeHandler {
```
Comment (xlab) [inline app/app.go:887]: Seems that this change is not related to wasmbindings

File: scripts/setup.sh [hunk 8]
```diff
@@ -5,6 +5,6 @@ test_mnemonic="wage thunder live sense resemble foil apple course spin horse gla
 
 persistenceCore init test --chain-id test
 echo $test_mnemonic | persistenceCore keys add test --recover --keyring-backend test
-persistenceCore add-genesis-account test 100000000000000stake --keyring-backend test
-persistenceCore gentx test 10000000stake --chain-id test --keyring-backend test
+persistenceCore add-genesis-account test 100000000000000uxprt --keyring-backend test
```
Comment (xlab) [inline scripts/setup.sh:8]: It can have both, especially when `stake` used  to delegate.

**Review by xlab — CHANGES_REQUESTED** [review]
Looks good, but this diff has some parts touched that are not related to wasmbindings

File: app/app.go (hunk 6 above)
Comment (Tikaryan) [reply app/app.go:577]: I don't think we are querying the `staking` module. So I removed it.

File: app/app.go (hunk 7 above)
Comment (Tikaryan) [reply app/app.go:887]: Yeah, it is not. It was an unused function.

File: scripts/setup.sh (hunk 8 above)
Comment (Tikaryan) [reply scripts/setup.sh:8]: So should I add both here? `100000000000000uxprt` and `100000000000000stake` ?

**Comment by Tikaryan** [comment]
I rebased the `development` branch with the latest `main`.
The Main branch had major changes so again had to add the changes to this PR.

@xlab Check once and let me know if everything looks fine.

File: scripts/setup.sh [hunk 9]
```diff
@@ -5,6 +5,6 @@ test_mnemonic="wage thunder live sense resemble foil apple course spin horse gla
 
 persistenceCore init test --chain-id test
 echo $test_mnemonic | persistenceCore keys add test --recover --keyring-backend test
-persistenceCore add-genesis-account test 100000000000000stake --keyring-backend test
-persistenceCore gentx test 10000000stake --chain-id test --keyring-backend test
+persistenceCore add-genesis-account test 100000000000000uxprt 100000000000000stake --keyring-backend test
```
Comment (xlab) [inline scripts/setup.sh:8]: must be `100000000000000uxprt,100000000000000stake`

```
Error: accepts 2 arg(s), received 3
Usage:
  persistenceCore add-genesis-account [address_or_key_name] [coin][,[coin]] [flags]
```

**Review by xlab — CHANGES_REQUESTED** [review]

File: scripts/setup.sh (hunk 9 above)
Comment (Tikaryan) [reply scripts/setup.sh]: Done.

File: scripts/setup.sh [hunk 10]
```diff
@@ -5,6 +5,6 @@ test_mnemonic="wage thunder live sense resemble foil apple course spin horse gla
 
 persistenceCore init test --chain-id test
 echo $test_mnemonic | persistenceCore keys add test --recover --keyring-backend test
-persistenceCore add-genesis-account test 100000000000000uxprt 100000000000000stake --keyring-backend test
-persistenceCore gentx test 10000000uxprt 10000000stake --chain-id test --keyring-backend test
+persistenceCore add-genesis-account test 100000000000000uxprt,100000000000000stake --keyring-backend test
+persistenceCore gentx test 10000000uxprt --chain-id test --keyring-backend test
```
Comment (xlab) [inline scripts/setup.sh:9]: gentx uses stake
Screenshot 2023-03-15 at 09 30 03

**Review by xlab — CHANGES_REQUESTED** [review]

File: scripts/setup.sh (hunk 10 above)
Comment (Tikaryan) [reply scripts/setup.sh]: Done.

**Review by xlab — APPROVED** [review]

**Comment by xlab** [comment]
Closes 4130961591
